homeassistant/components/zwave_js/api.py

```diff
@@ -183,6 +243,56 @@ async def async_handle_failed_command_func(
     return async_handle_failed_command_func
 
 
+def get_provisioning_info(
+    msg: dict,
+) -> str | ProvisioningEntry | QRProvisioningInformation | None:
+    """Process provisioning parameters and return the appropriate value."""
+    if planned_provisioning_entry := msg.get(PLANNED_PROVISIIONING_ENTRY):
+        return ProvisioningEntry(
+            dsk=planned_provisioning_entry[DSK],
+            security_classes=[
+                SecurityClass(sec_cls)
+                for sec_cls in planned_provisioning_entry[SECURITY_CLASSES]
+            ],
+            additional_properties={
+                k: v
+                for k, v in planned_provisioning_entry.items()
+                if k not in (DSK, SECURITY_CLASSES)
+            },
+        )
+
+    if qr_provisioning_information := msg.get(QR_PROVISIONING_INFORMATION):
+        protocols = [
+            Protocols(proto)
+            for proto in qr_provisioning_information.get(SUPPORTED_PROTOCOLS, [])
+        ]
+        return QRProvisioningInformation(
+            version=QRCodeVersion(qr_provisioning_information[VERSION]),
+            security_classes=[
+                SecurityClass(sec_cls)
+                for sec_cls in qr_provisioning_information[SECURITY_CLASSES]
+            ],
+            dsk=qr_provisioning_information[DSK],
+            generic_device_class=qr_provisioning_information[GENERIC_DEVICE_CLASS],
+            specific_device_class=qr_provisioning_information[SPECIFIC_DEVICE_CLASS],
+            installer_icon_type=qr_provisioning_information[INSTALLER_ICON_TYPE],
+            manufacturer_id=qr_provisioning_information[MANUFACTURER_ID],
+            product_type=qr_provisioning_information[PRODUCT_TYPE],
+            product_id=qr_provisioning_information[PRODUCT_ID],
+            application_version=qr_provisioning_information[APPLICATION_VERSION],
+            max_inclusion_request_interval=qr_provisioning_information.get(
+                MAX_INCLUSION_REQUEST_INTERVAL
+            ),
+            uuid=qr_provisioning_information.get(UUID),
+            supported_protocols=protocols if protocols else None,
+        )
+
+    if qr_code_string := msg.get(QR_CODE_STRING):
+        return qr_code_string
+
+    return None
```

Comment:
This case should not be able to happen if we validate the input first with `cv.has_at_least_one_key`. We can `raise RuntimeError` here.

Reply:
`add_node` and `replace_failed_node` do not require provisioning information ahead of time so this is still valid for those commands.

Reply:
I don't really like calling functions with all optional arguments and returning None if no input is passed.

Maybe we can make one or more validator functions instead that are run during input validation and return the correct provisioning or string?

Reply:
Looks possible at least if we break out the inclusion strategy validation.

Reply:
Can we add a feature to the core schema decorator to allow for additional validators? If so then we can do all of this via input validation

Reply:
Maybe in a separate PR, yes. Run it by Paulus.

Maybe allow a voluptuous schema besides the dict for the schema parameter in the decorator and have two different flows with the same result.

Reply:
https://github.com/home-assistant/core/pull/59856 waiting to hear back